### PR TITLE
Added GetConfigEnum and SetConfigEnum to ConfigSection

### PR DESCRIPTION
**Pull request by MarcasRealAccount**
Added support for semi reading and writing enums.
Made GetConfig work with sub sections by using '.' character.

### CODE REVIEW

File: src/Engine/Input/InputHandler.cpp [hunk 1]
```diff
@@ -78,24 +82,166 @@ namespace gp1 {
...
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouseRight } }, "Mouse.Right" },
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouseMiddle } }, "Mouse.Middle" },
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouse1 } }, "Mouse.1" },
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouse2 } }, "Mouse.2" },
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouse3 } }, "Mouse.3" },
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouse4 } }, "Mouse.4" },
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouse5 } }, "Mouse.5" },
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouse6 } }, "Mouse.6" },
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouse7 } }, "Mouse.7" },
+			{ { InputType::BUTTON, { InputLocation::MOUSE, buttons::mouse8 } }, "Mouse.8" },
```
Comment (Gaztin) [inline src/Engine/Input/InputHandler.cpp:238]: Aren't there only 5 mouse buttons? left, right, middle, extra-1 and extra-2?

Comment (MarcasRealAccount) [reply src/Engine/Input/InputHandler.cpp:238]: According to GLFW there's 8 mouse buttons.
Even though most mouses only support 5 buttons.

**Comment by Gaztin** [comment]
What are the practical advantages of using this over just having an "EnumStringify"-function with a big switch-case for all the buttons etc?

**Comment by MarcasRealAccount** [comment]
Just wait with adding this pr until I have added stuff for the controllers as well (since you added that first).

File: src/Engine/Input/InputHandler.cpp (hunk 1 above)
Comment (Gaztin) [reply src/Engine/Input/InputHandler.cpp:238]: Ah, okay!

**Comment by MarcasRealAccount** [comment]
I dont really know. Probably so that the enums can be decided at runtime too.

**Comment by MarcasRealAccount** [comment]
There we go, it should support Gamepad buttons and axises too now 😄

**Comment by Gaztin** [comment]
> 
> 
> I dont really know. Probably so that the enums can be decided at runtime too.

Yeah, question is if that is ever going to be needed, or if this is a little over-engineered in that sense? Just a thought.

**Comment by MarcasRealAccount** [comment]
Currently no the runtime ability isn't used anywhere, but you never know what will happen in the future.
But if you really want it to use a function instead for getting the name of an enum value, then I guess I might change it to do that instead.
In that case should I just make it take a function pointer or a std::function?

**Comment by Gaztin** [comment]
I think you misunderstand me. I'm wondering why we need to tell `GetConfigEnum` which enums there are, when the enum could just as well be stored in the input system itself.

I haven't taken a deep look at this input binding system so you'll have to excuse me, but I feel like it's very convoluted for what it does. In the end, all I wish for is that this system is intuitive for people to use.

**Comment by MarcasRealAccount** [comment]
So the `GetConfigEnum` function takes a `std::vector<std::pair<T, std::string>>` as the enum values because the way it works is that it reads the config as usual, after getting the string value of the config it goes through the list of enum values and finds the correct one. And returns the `T` value from that `std::pair`, such that it's possible to combine the index and location values from the config into a more intuitive way of formatting the config.

So the config file would look more like this:
```ini
[audioInput]
playWav = Key.1
playMp3 = Key.2
playFlac = Key.3

[inMenu]
closeMenu = Gamepad.Y

[onFoot]
openMenu = Gamepad.Y
lookX = Gamepad.LeftTrigger
lookY = Gamepad.RightTrigger
```
Instead of this:
```ini
[audioInput.playWav]
index = 49
location = 0

[audioInput.playMp3]
index = 50
location = 0

[audioInput.playFlac]
index = 51
location = 0

[inMenu.closeMenu]
index = 3
location = 2

[onFoot.openMenu]
index = 3
location = 2

[onFoot.lookX]
index = 4
location = 2

[onFoot.lookY]
index = 5
location = 2
```

And the enums are stored in the input system's `InputHandler.cpp` file at the bottom.
Right under the very amazing comment `//PLEASE GOD NOOOO:`

**Comment by Gaztin** [comment]
Ah, I see. So the "enums" are not known at compile-time?

**Comment by MarcasRealAccount** [comment]
No they are known at runtime.
Since C++ still have no proper way to stringify an enum.

**Comment by Gaztin** [comment]
Okay, so back to my original question. Why can't we have a function that stringifies the enums? A function that can be globally accessible from the `GetConfigEnum` function so that we don't have to pass the vector.

**Comment by Gaztin** [comment]
Or rather, why can't the `InputNames` variable be accessed from the `GetConfigEnum` function globally?

**Comment by MarcasRealAccount** [comment]
So this is the functions inside the `ConfigSection.h` file if you need it to understand some things.
```cpp
// Sets a config value with an enum value name.
template <typename T>
void SetConfigEnum(const std::string& key, T value, const std::vector<std::pair<T, std::string>>& enumNames) {
  auto itr = enumNames.begin();
  while (itr != enumNames.end()) {
    if (itr->first == value) {
      SetConfig(key, itr->second);
      break;
    }
    itr++;
  }
}

// Gets the element T that the config's value was the same as.
template <typename T>
T GetConfigEnum(const std::string& key, T def, const std::vector<std::pair<T, std::string>>& enumNames) {
  std::string defName = "";
  auto itr = enumNames.begin();
  while (itr != enumNames.end()) {
    if (itr->first == def) {
      defName = itr->second;
      break;
    }
    itr++;
  }
  if (itr == enumNames.end()) return def;

  const std::string& configValue = GetConfig(key, defName);

  itr = enumNames.begin();
  while (itr != enumNames.end()) {
    if (itr->second == configValue)
      return itr->first;
    itr++;
  }
  return def;
}
```

As you can probably see it requires all the names of all the enum values to figure out what the actual return value should be in the `GetConfigEnum`. For the `SetConfigEnum` having an enum stringify function would work pretty ok, but it wouldn't really work for the getter though.

And I believe that the Config System shouldn't really know that the Input System exists, because if anyone wants to extract the Config System for their own projects then it would work in their own project with little to no errors.

**Comment by Gaztin** [comment]
All right. I'll try and adapt the `GetConfigEnum` to try and visualize my solution:
```cpp
// Gets the element T that the config's value was the same as.
template <typename T>
T GetConfigEnum(const std::string& key, T def) {
  const std::string defName = FindDefName(def);
  if (defName.empty()) return def;

  const std::string& configValue = GetConfig(key, defName);
  return FindConfigValue(configValue);
}
```
In above example, `FindDefName` and `FindConfigValue` are global functions that performs a switch on a defined set of values.

Granted, I thought this system was for the input system and the input system *only*. If you want this to be generic, then I understand why you need to pass the look-up table to the function. But if we're calling these functions a lot each frame then it might be nicer to the cache.

**Comment by MarcasRealAccount** [comment]
I do understand that it's not great for doing it each frame, but it's a config system and it should only ever be used a few times.
It's not a system that should ever run every frame, due to writing to a file for instance.

Oh no the system should be available to anything really.
I mean it's also in use for shaders too, since I wanted the shaders to be read and parsed without writing code for a specific shader it uses the config system to store the shader's attributes and uniforms so if you look at `src/Game/Assets/Shaders/debugShader.ini` that config file holds all the attributes and uniforms that the DebugShader uses.

And without the `GetConfigEnum` function it would be very difficult to write the shader config files.
Due to having to figure out the enums index which isn't that intuitive, instead it would now use some generic names for the different uniform types.

Though I don't really know if the names I chose are the best, but here's the list so you can give me feedback on them:
```cpp
std::vector<std::pair<UniformType, std::string>> UniformTypeNames = {
  { UniformType::FLOAT, "Float" },
  { UniformType::FLOAT_VEC2, "FVec2" },
  { UniformType::FLOAT_VEC3, "FVec3" },
  { UniformType::FLOAT_VEC4, "FVec4" },
  { UniformType::INT, "Int" },
  { UniformType::INT_VEC2, "IVec2" },
  { UniformType::INT_VEC3, "IVec3" },
  { UniformType::INT_VEC4, "IVec4" },
  { UniformType::UINT, "UInt" },
  { UniformType::UINT_VEC2, "UVec2" },
  { UniformType::UINT_VEC3, "UVec3" },
  { UniformType::UINT_VEC4, "UVec4" },
  { UniformType::FLOAT_MAT2, "FMat2" },
  { UniformType::FLOAT_MAT3, "FMat3" },
  { UniformType::FLOAT_MAT4, "FMat4" }
};
```

**Comment by Gaztin** [comment]
Ah, okay! Then all is good here. Sorry for the confusion. I feel well informed now 😊

File: include/Engine/Utility/Config/ConfigSection.h [hunk 2]
```diff
@@ -40,6 +41,43 @@ namespace gp1 {
...
+						break;
+					}
+					itr++;
+				}
+			}
+
+			// Gets the element T that the config's value was the same as.
+			template <typename T>
+			T GetConfigEnum(const std::string& key, T def, const std::vector<std::pair<T, std::string>>& enumNames) {
+				std::string defName = "";
```
Comment (Gaztin) [inline include/Engine/Utility/Config/ConfigSection.h:60]: Just using `itr->second` directly should work just as well, right? No need for a temporary copy!

File: include/Engine/Utility/Config/ConfigSection.h [hunk 3]
```diff
@@ -40,6 +41,43 @@ namespace gp1 {
 			// Get T value of a config.
 			template <typename T>
 			T GetConfigTyped(const std::string& key, T def);
+			// Sets a config value with an enum value name.
+			template <typename T>
+			void SetConfigEnum(const std::string& key, T value, const std::vector<std::pair<T, std::string>>& enumNames) {
```
Comment (Gaztin) [inline include/Engine/Utility/Config/ConfigSection.h:46]: Maybe we can make this a bit shorter by adding an alias? Something like
```cpp
template <typename T> using EnumVector = std::vector<std::pair<T, std::string>>;
```
that we later can use for `InputNames` etc!

**Review by Gaztin — CHANGES_REQUESTED** [review]

File: include/Engine/Utility/Config/ConfigSection.h (hunk 2 above)
Comment (MarcasRealAccount) [reply include/Engine/Utility/Config/ConfigSection.h:60]: Maybe I guess I'll try to remove the temp copy.

File: include/Engine/Utility/Config/ConfigSection.h (hunk 3 above)
Comment (MarcasRealAccount) [reply include/Engine/Utility/Config/ConfigSection.h:46]: Yeah that sounds good. 😄

**Pull request merged**